internal/controller/stages/regular_stages.go

```diff
@@ -0,0 +1,1981 @@
+	}
+
+	// Return the reconcile error if it exists.
+	if reconcileErr != nil {
+		return ctrl.Result{}, reconcileErr
+	}
+	// Immediate requeue if needed.
+	if needsRequeue {
+		return ctrl.Result{Requeue: true}, nil
+	}
+	// Otherwise, requeue after a delay.
+	// TODO: Make the requeue delay configurable.
+	return ctrl.Result{RequeueAfter: 5 * time.Minute}, nil
+}
+
+func (r *RegularStagesReconciler) reconcile(
+	ctx context.Context,
+	stage *kargoapi.Stage,
+	startTime time.Time,
+) (kargoapi.StageStatus, bool, error) {
+	logger := logging.LoggerFromContext(ctx)
+	newStatus := *stage.Status.DeepCopy()
+
+	// Mark the Stage as reconciling.
+	conditions.Set(&newStatus, &metav1.Condition{
+		Type:               kargoapi.ConditionTypeReconciling,
+		Status:             metav1.ConditionTrue,
+		Reason:             "Reconciling",
+		ObservedGeneration: stage.Generation,
+	})
+
+	var requestRequeue bool
+	subReconcilers := []struct {
+		name      string
+		reconcile func() (kargoapi.StageStatus, error)
+	}{
+		{
+			name: "syncing Promotions",
+			reconcile: func() (kargoapi.StageStatus, error) {
+				status, hasPendingPromotions, err := r.syncPromotions(ctx, stage)
+				if err != nil {
+					err = fmt.Errorf("failed to sync Promotions: %w", err)
+				}
+				if status.CurrentPromotion == nil && hasPendingPromotions {
+					requestRequeue = true
+				}
+				return status, err
+			},
+		},
+		{
+			name: "assessing health",
+			reconcile: func() (kargoapi.StageStatus, error) {
+				return r.assessHealth(ctx, stage), nil
+			},
+		},
+		{
+			name: "verifying Stage Freight",
+			reconcile: func() (kargoapi.StageStatus, error) {
+				status, err := r.verifyStageFreight(ctx, stage, startTime, time.Now)
+				if err != nil {
+					err = fmt.Errorf("failed to verify Stage Freight: %w", err)
+				}
+				return status, err
+			},
+		},
+		{
+			name: "verifying Freight for Stage",
+			reconcile: func() (kargoapi.StageStatus, error) {
+				status, err := r.verifyFreightForStage(ctx, stage)
+				if err != nil {
+					err = fmt.Errorf("failed to verify Freight for Stage: %w", err)
+				}
+				return status, err
+			},
+		},
+		{
+			name: "auto-promoting Freight",
+			reconcile: func() (kargoapi.StageStatus, error) {
+				status, err := r.autoPromoteFreight(ctx, stage)
+				if err != nil {
+					err = fmt.Errorf("failed to auto-promote Freight: %w", err)
+				}
+				return status, err
+			},
+		},
+	}
+	for _, subR := range subReconcilers {
+		logger.Debug(subR.name)
+
+		// Reconcile the Stage with the sub-reconciler.
+		var err error
+		newStatus, err = subR.reconcile()
+
+		// Summarize the conditions after each sub-reconciler to ensure that
+		// we have a consistent view of the Stage status.
+		summarizeConditions(stage, &newStatus, err)
+
+		// If an error occurred during the sub-reconciler, then we should
+		// return the error and request an immediate requeue.
+		if err != nil {
+			return newStatus, false, err
```

Comment:
I'm pretty sure you've left requeue as false because a non-nil error will cause a requeue anyway, but to the casual reader, this may appear to contradict the comment above. wdyt about either setting it to true _or_ adding a comment that a non-nil error will trigger a requeue?